### PR TITLE
Respect maxSize in TableChangesSplitSource.getNextBatch

### PR DESCRIPTION
**Pull request by homar**
## Description
Fixes: https://github.com/trinodb/trino/issues/17182

## Release notes

(x) This is not user-visible or docs only and no release notes are required.
() Release notes are required, please propose a release note for me.
() Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesSplitSource.java [hunk 1]
```diff
@@ -49,84 +51,95 @@
...
-        catch (IOException e) {
-            throw new TrinoException(DELTA_LAKE_FILESYSTEM_ERROR, "Failed to access table metadata", e);
+                    catch (IOException e) {
+                        throw new TrinoException(DELTA_LAKE_FILESYSTEM_ERROR, "Failed to access table metadata", e);
+                    }
+                });
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesSplitSource.java:134]: Just for the understanding of the other reviewers: `maxSize` comes from `query.schedule-split-batch-size` setting.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesSplitSource.java [hunk 2]
```diff
@@ -49,84 +51,95 @@
...
-                        splits.add(mapToDeltaLakeTableChangesSplit(
-                                commitInfo,
-                                DATA_FILE,
-                                addEntry.getSize(),
-                                addEntry.getPath(),
-                                addEntry.getCanonicalPartitionValues()));
+        return LongStream.range(currentVersion, tableReadVersion + 1)
+                .boxed()
+                .flatMap(version -> {
+                    try {
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesSplitSource.java:76]: nit: Extract the anonymous code to a method.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesSplitSource.java [hunk 3]
```diff
@@ -49,84 +51,95 @@
...
             TrinoFileSystemFactory fileSystemFactory,
             TableChangesTableFunctionHandle functionHandle)
     {
         tableLocation = functionHandle.tableLocation();
-        fileSystem = fileSystemFactory.create(session);
-        currentVersion = new AtomicLong(functionHandle.firstReadVersion());
-        tableReadVersion = functionHandle.tableReadVersion();
-        transactionLogDir = getTransactionLogDir(tableLocation);
+        splits = prepareSplits(
+                functionHandle.firstReadVersion(),
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesSplitSource.java:64]: Can we have a test similar to `io.trino.plugin.deltalake.TestDeltaLakeConnectorTest#testReadCdfChanges` in the context of dealing with 

```
"query.schedule-split-batch-size", "1",
```

see `io.trino.plugin.hive.TestBucketedQueryWithManySplits`

With the default value of `1000` that we have by default for `query.schedule-split-batch-size`  I guess we're not currently actually testing this functionality with the current codebase (and your PR doesn't contain a corresponding test).

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesSplitSource.java [hunk 4]
```diff
@@ -49,84 +51,95 @@
...
+                    catch (IOException e) {
+                        throw new TrinoException(DELTA_LAKE_FILESYSTEM_ERROR, "Failed to access table metadata", e);
+                    }
+                });
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+    {
+        List<ConnectorSplit> result = new ArrayList<>();
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesSplitSource.java:136]: nit: use `ImmutableList.Builder`

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestCdfWithManySplits.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        catch (Throwable e) {
+            closeAllSuppress(e, queryRunner);
+            throw e;
+        }
+
+        return queryRunner;
+    }
+
+    @Test
+    public void testReadCdfChangesWithMaxSplitsSetToOne()
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestCdfWithManySplits.java:85]: `testReadCdfChangesWithMaxSplitsSetToOne` -> `testReadCdfChanges`

the fact that the max splits is set to 1 is actually handled in the query runner builder

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestCdfWithManySplits.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    {
+        String tableName = "test_basic_operations_on_table_with_cdf_enabled_" + randomNameSuffix();
+        assertUpdate("CREATE TABLE " + tableName + " (page_url VARCHAR, domain VARCHAR, views INTEGER) WITH (change_data_feed_enabled = true)");
+        assertUpdate("INSERT INTO " + tableName + " VALUES('url1', 'domain1', 1), ('url2', 'domain2', 2), ('url3', 'domain3', 3)", 3);
+        assertUpdate("INSERT INTO " + tableName + " VALUES('url4', 'domain4', 4), ('url5', 'domain5', 2), ('url6', 'domain6', 6)", 3);
+
+        assertUpdate("UPDATE " + tableName + " SET page_url = 'url22' WHERE views = 2", 2);
+        assertTableChangesQuery("SELECT * FROM TABLE(system.table_changes('test_schema', '" + tableName + "'))",
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestCdfWithManySplits.java:93]:
```
trino:tiny> select * from table(system.table_changes('tiny', 'test1'));
 page_url | domain  | views |   _change_type   | _commit_version |      _commit_timestamp      
----------+---------+-------+------------------+-----------------+-----------------------------
 url1     | domain1 |     1 | insert           |               1 | 2023-07-21 08:35:51.846 UTC 
 url2     | domain2 |     2 | insert           |               1 | 2023-07-21 08:35:51.846 UTC 
 url3     | domain3 |     3 | insert           |               1 | 2023-07-21 08:35:51.846 UTC 
 url4     | domain4 |     4 | insert           |               2 | 2023-07-21 08:36:05.199 UTC 
 url5     | domain5 |     2 | insert           |               2 | 2023-07-21 08:36:05.199 UTC 
 url6     | domain6 |     6 | insert           |               2 | 2023-07-21 08:36:05.199 UTC 
 url22    | domain5 |     2 | update_postimage |               3 | 2023-07-21 08:36:20.594 UTC 
 url22    | domain2 |     2 | update_postimage |               3 | 2023-07-21 08:36:20.594 UTC 
 url2     | domain2 |     2 | update_preimage  |               3 | 2023-07-21 08:36:20.594 UTC 
 url5     | domain5 |     2 | update_preimage  |               3 | 2023-07-21 08:36:20.594 UTC 
(10 rows)

Query 20230721_084139_00016_jkxp6, FINISHED, 1 node
http://localhost:8080/ui/query.html?20230721_084139_00016_jkxp6
Splits: 3 total, 3 done (100.00%)
CPU Time: 0.0s total,     0 rows/s,     0B/s, 72% active
Per Node: 0.0 parallelism,     0 rows/s,     0B/s
Parallelism: 0.0
Peak Memory: 1.21KB
1.10 [0 rows, 0B] [0 rows/s, 0B/s]

trino:tiny> explain analyze select * from table(system.table_changes('tiny', 'test1'));
                                                                              Query Plan                                                                               
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Trino version: dev                                                                                                                                                    
 Queued: 255.83us, Analysis: 50.45ms, Planning: 9.71ms, Execution: 695.27ms                                                                                            
 Fragment 1 [SOURCE]                                                                                                                                                   
     CPU: 26.54ms, Scheduled: 39.07ms, Blocked 0.00ns (Input: 0.00ns, Output: 0.00ns), Input: 0 rows (0B); per task: avg.: 0.00 std.dev.: 0.00, Output: 10 rows (526B) 
     Output layout: [page_url, domain, views, _change_type, _commit_version, _commit_timestamp]                                                                        
     Output partitioning: SINGLE []                                                                                                                                    
     TableFunctionProcessor[name = table_changes, properOutputs = [page_url, domain, views, _change_type, _commit_version, _commit_timestamp]]                         
         Layout: [page_url:varchar, domain:varchar, views:integer, _change_type:varchar, _commit_version:bigint, _commit_timestamp:timestamp(3) with time zone]        
         CPU: 27.00ms (100.00%), Scheduled: 38.00ms (100.00%), Blocked: 0.00ns (?%), Output: 10 rows (526B)                                                            
         Input avg.: 0.00 rows, Input std.dev.: ?%                                                                                                                     
                                                                                                                                                                       
                                                                                                                                                                       
(1 row)
```
I find it a bit strange that the `TableFunctionProcessor` doesn't return the right amount of rows in `EXPLAIN ANALYZE`. Ideas?

**Review by findinpath — APPROVED** [review]

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestCdfWithManySplits.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+import static io.trino.plugin.base.util.Closables.closeAllSuppress;
+import static io.trino.plugin.deltalake.DeltaLakeQueryRunner.DELTA_CATALOG;
+import static io.trino.testing.TestingNames.randomNameSuffix;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static io.trino.testing.containers.Minio.MINIO_ACCESS_KEY;
+import static io.trino.testing.containers.Minio.MINIO_SECRET_KEY;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestCdfWithManySplits
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestCdfWithManySplits.java:33]: I think the total split in the current test isn't "many". How about populating more rows? 
We could use aggreagtion for simplifying the assertion.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestCdfWithManySplits.java [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+{
+    private static final String SCHEMA = "test_schema";
+
+    private static final String bucketName = "test-bucket-" + randomNameSuffix();
+
+    @Override
+    protected QueryRunner createQueryRunner()
+            throws Exception
+    {
+        Minio minio = closeAfterClass(Minio.builder().build());
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestCdfWithManySplits.java:44]: Why not use local file system?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestCdfWithManySplits.java [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        minio.start();
+        minio.createBucket(bucketName);
+
+        DistributedQueryRunner queryRunner = DistributedQueryRunner.builder(testSessionBuilder()
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestCdfWithManySplits.java:48]: We could use `DeltaLakeQueryRunner` instead.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestCdfWithManySplits.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+            queryRunner.createCatalog(DELTA_CATALOG, DeltaLakeConnectorFactory.CONNECTOR_NAME, ImmutableMap.<String, String>builder()
+                    .put("hive.metastore", "file")
+                    .put("hive.metastore.catalog.dir", queryRunner.getCoordinator().getBaseDataDir().resolve("file-metastore").toString())
+                    .put("hive.metastore.disable-location-checks", "true")
+                    .put("hive.s3.aws-access-key", MINIO_ACCESS_KEY)
+                    .put("hive.s3.aws-secret-key", MINIO_SECRET_KEY)
+                    .put("hive.s3.endpoint", minio.getMinioAddress())
+                    .put("hive.s3.path-style-access", "true")
+                    .put("delta.enable-non-concurrent-writes", "true")
+                    .put("delta.register-table-procedure.enabled", "true")
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestCdfWithManySplits.java:71]: `delta.register-table-procedure.enabled` looks redundant.

**Review by ebyhr — APPROVED** [review]

**Pull request merged**